CorsixTH/Lua/humanoid_actions/seek_room.lua

```diff
@@ -246,115 +253,162 @@ local function action_seek_room_start(action, humanoid)
+    action.must_happen = true
 
-      local remove_callback = --[[persistable:action_seek_room_remove_callback]] function(rm)
-        if rm.room_info.id == "research" then
-          humanoid:updateMessage("research")
-        end
-      end -- End of remove_callback function
-      action.remove_callback = remove_callback
-      humanoid:registerRoomRemoveCallback(remove_callback)
+    local remove_callback = --[[persistable:action_seek_room_remove_callback]] function(rm)
+      humanoid:updateMessage("research")
+    end -- End of remove_callback function
+    action.remove_callback = remove_callback
+    humanoid:registerRoomRemoveCallback(remove_callback)
+
+    local build_callback
 
-      local build_callback
-      build_callback = --[[persistable:action_seek_room_build_callback]] function(rm)
-        -- if research room was built, message may need to be updated
-        if rm.room_info.id == "research" then
-          humanoid:updateMessage("research")
+    local staff_callback
+    staff_callback = --[[persistable:action_seek_room_staff_callback]] function(staff)
+      -- we might have hired or fired a staff member
+      -- technically we don't care about Receptionist or Handyman
+      if staff then
+        if staff.humanoid_class == "Receptionist" or staff.humanoid_class == "Handyman" then
+          return
         end
+      end
+      -- update the message either way
+      humanoid:updateMessage("research")
 
-        local found = false
-        if rm.room_info.id == action.room_type then
-          found = true
-        elseif rm.room_info.id == action.room_type_needed then
-          -- So the room that we're going to is not actually the room we waited for to be built.
-          -- Example: Will go to ward, but is waiting for the operating theatre.
-          -- Clean up and start over to find the room we actually want to go to.
+      local room_req = humanoid.hospital:checkDiseaseRequirements(humanoid.disease.id)
+      if not room_req then
+        local room = action_seek_room_find_room(action, humanoid)
+        if room then
           TheApp.ui.bottom_panel:removeMessage(humanoid)
           humanoid:unregisterRoomBuildCallback(build_callback)
           humanoid:unregisterRoomRemoveCallback(remove_callback)
-          action.room_type_needed = nil
-          action_seek_room_start(action, humanoid)
-        elseif not humanoid.diagnosed then
-          -- Waiting for a diagnosis room, we need to go through the list - unless it is gp
-          if action.room_type ~= "gp" then
-            for i = 1, #humanoid.available_diagnosis_rooms do
-              if humanoid.available_diagnosis_rooms[i].id == rm.room_info.id then
-                found = true
-              end
+          humanoid:unregisterStaffCallback(staff_callback)
+          action_seek_room_goto_room(room, humanoid, action.diagnosis_room)
+        end
+      end
+    end -- End of staff_callback function
+    action.staff_callback = staff_callback
```

Comment:
Callback for when a staff member is hired/fired to update the message will also enable the Research button on the fax, not just the change in message